### DISCUSSION TITLE
Private packages (artifactory) not found via build.sbt

### DISCUSSION THREAD

**Discussion by eamonnsullivan** [renovatebot/renovate#32848] [Request Help]
### How are you running Renovate?

A Mend.io-hosted app

### Please tell us more about your question or problem

We use Renovate for a lot of our npm repos. I'm trying to onboard a Scala backend service that uses `build.sbt`. The public dependencies (via public Maven and typesafe) are found without issues and the onboarding PR says several pull requests will be raised immediately against dependencies like Scala runtimes, AWS libs, etc. However, it can't find our private packages, like the first two listed here (I've anonymised it a bit, where xxx is my organisation's name): 
```
libraryDependencies ++= Seq(
  // core
  "xxx.shared" %% "xxx-scalatra-chassis" % "X.X",
  "xxx.shared" %% "xxx-software-domain" % "X.X",
  "commons-io" % "commons-io" % "2.6",
```
In the pull request, I see the error: `Failed to look up sbt-package package xxx.shared:xxx-scalatra-chassis_2.13`. 

The bottom of my build.sbt has the following: 
```
resolvers ++= Seq(
  "Local Ivy repository" at "file://" + Path.userHome + "/.ivy2/local",
  "XXX Maven Releases" at "https://dev.xxx.co.uk/maven2/releases/",
  "XXX Maven Snapshots" at "https://dev.xxx.co.uk/maven2/snapshots",
  "XXX Artifactory" at "https://dev.xxx.co.uk/artifactory/repo",
  "Typesafe repository" at "https://repo.typesafe.com/typesafe/releases/"
)
```

My renovate.json pulls mostly from a central config, which all seems to be picked up (judging by the description in the onboarding PR) and I've tried adding some hostRules at the end to pick up our artifactory credentials. The secrets are set, but I still get the same errors about being unable to look up those private deps. 

```json
{
  "extends": [
    "github>xxx/configs-core//renovate-config/default"
  ],
  "hostRules": [
    {
      "hostType": "maven",
      "matchHost": "https://dev.xxx.co.uk",
      "username": "{{ secrets.ARTIFACTORY_USERNAME }}",
      "password": "{{ secrets.ARTIFACTORY_PASSWORD}}"
    }
  ],
  "packageRules": [
    {
      "matchDatasources": ["sbt-package"],
      "matchPackagePrefixes": ["xxx.shared"],
      "registryUrls": [
        "https://dev.xxx.co.uk/maven2/releases/",
        "https://dev.xxx.co.uk/maven2/snapshots",
        "https://dev.xxx.co.uk/artifactory/repo"
      ]
    }
  ]
}
```

I see TLS errors in the logs. I think the issue may be because we use mTLS to authenticate with our artifactory. (We use username and password as well, for publishing.)

### Logs (if relevant)

<details><summary>Logs</summary>

```
{
  "failedUrl": "https://artifactory.dev.bbc.co.uk/artifactory/int-bbc-releases/bbc/shared/bbc.shared/"
  "err": {
    "name": "RequestError",
    "code": "EPROTO",
    "timings": {
      "start": 1733167223765,
      "socket": 1733167223765,
      "lookup": 1733167223780,
      "connect": 1733167223871,
      "upload": 1733167225899,
      "error": 1733167225899,
      "phases": {
        "wait": 0,
        "dns": 15,
        "tcp": 91,
        "request": 2028,
        "total": 2134
      }
    },
    "message": "write EPROTO C0FC8232AA7F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1605:SSL alert number 40\n",
    "stack": "RequestError: write EPROTO C0FC8232AA7F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1605:SSL alert number 40\n\n    at ClientRequest.<anonymous> (/usr/local/renovate/node_modules/.pnpm/got@11.8.6/node_modules/got/dist/source/core/index.js:970:111)\n    at Object.onceWrapper (node:events:633:26)\n    at ClientRequest.emit (node:events:530:35)\n    at ClientRequest.emit (node:domain:489:12)\n    at ClientRequest.origin.emit (/usr/local/renovate/node_modules/.pnpm/@szmarczak+http-timer@4.0.6/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at emitErrorEvent (node:_http_client:103:11)\n    at TLSSocket.socketErrorListener (node:_http_client:506:5)\n    at TLSSocket.emit (node:events:518:28)\n    at TLSSocket.emit (node:domain:489:12)\n    at emitErrorNT (node:internal/streams/destroy:170:8)\n    at emitErrorCloseNT (node:internal/streams/destroy:129:3)\n    at processTicksAndRejections (node:internal/process/task_queues:90:21)\n    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:87:19)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/39.42.4 (https://github.com/renovatebot/renovate)",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://artifactory.dev.bbc.co.uk/artifactory/int-bbc-releases/bbc/shared/bbc.shared/",
      "hostType": "sbt",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    }
  }
}
```

</details>

**Comment by eamonnsullivan** [comment]
I've updated the initial post, with logs. I think I know what the issue is: We use mTLS for connecting to artifactory. It works on Jenkins, Codebuild and Github Actions because we can inject a certificate. Is there a way to do that in Renovate?

**Reply by Churro** [reply]
Yes, there is. See `httpsCertificate` ([here](https://docs.renovatebot.com/configuration-options/#httpscertificate)) and related options.

**Reply by eamonnsullivan** [reply]
Our build system generates short-lived certs (an hour or so) for this kind of thing, so I'm not sure how I'd get these to work. Will need a think.

**Reply by rarkins** [reply]
If self-hosting, you could do this as part of `config.js` dynamically. For Mend-hosting, it awaits a feature we're working on to allow secrets to be uploaded via API. You'd need a process to generate the certificates periodically and push a new value via Mend's API, to be used in jobs.

**Reply by eamonnsullivan** [reply]
That sounds ideal. Thank you! Is there an issue I can subscribe to? 

I guess we can enable this now to handle our public dependencies and await this feature for the internal deps.

**Reply by rarkins** [reply]
No, we mostly only have issues in this repo for the open source. The feature I mentioned is a Mend commercial hosting feature
